### PR TITLE
Use max_primary_shard_size instead of max_size

### PR DESCRIPTION
**Pull request by ph**
Based on change from https://github.com/elastic/elasticsearch/pull/69995

Users are having a hard time to have the "right sized shards" (~50gb),
but they also need to adjust the number of primaries to keep up with the
ingest load. This doesn't well with the `max_size` option but it should
work nicely with the newly introduced `max_primary_shard_size`.

This change the default ILM policy shipped with beats.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by ph** [comment]
@joegallo I took care of the default ILM policy shipped with beats.

**Comment by joegallo** [comment]
Possibly a bit of a BWC concern here: `max_primary_shard_size` exists in 7.13 and later -- will this potentially end up talking to a pre-7.13 Elasticsearch? If so, unfortunately, I suppose we'd have to have some kind of a version checking. (More of a concern on the 7.x version of this PR than the 8.0 version.)

**Comment by ph** [comment]
@joegallo Really good point about BWC, I am reluctant a bit to add version check for only a small changes and add the additional tech deb. If we remove the settings in 8.0 I think we should only target this change for 8.0 and not make it land in 7.x. @urso

**Comment by ruflin** [comment]
When exactly is this ILM policy setup? Does the user have an option to opt out? 

"Official" the Elasticsearch should always be newer then the Beat, unfortunately we never enforce it. We do now with Elastic Agent. So I would argue this change is ok.

@jpountz Would be good to also get your take on this change. Should we make the same for the Elasticsearch default ilm policies if not already done?

**Comment by jpountz** [comment]
> the Elasticsearch should always be newer then the Beat [...] We do now with Elastic Agent. So I would argue this change is ok.

Strong :+1:. It would be too hard for Agent to rely on the newer Elasticsearch features otherwise.

> @jpountz Would be good to also get your take on this change. Should we make the same for the Elasticsearch default ilm policies if not already done?

This is actually a question that we are currently trying to figure out as this has interesting connections with how autoscaling makes the number of shards dynamic and whether or not a shrink action is configured. @dakrone What would you advise in the short term, should Beats/Agent hold on and keep using `max_size` until we figured things out, or should they move to `max_primary_shard_size` as this PR suggests?

**Comment by dakrone** [comment]
> Would be good to also get your take on this change. Should we make the same for the Elasticsearch default ilm policies if not already done?

This is already done in https://github.com/elastic/elasticsearch/pull/69995

> @dakrone What would you advise in the short term, should Beats/Agent hold on and keep using max_size until we figured things out, or should they move to max_primary_shard_size as this PR suggests?

They should move to `max_primary_shard_size`. The more we can make independent of scaling knobs (such as the number of primary shards in this case) the better I think.

**Comment by urso** [comment]
> I am reluctant a bit to add version check for only a small changes and add the additional tech deb.

For settings in templates or ILM policy it is common practice if the setting improves user experience/stability. I wouldn't consider it technical debt here. The code is supposed to work with ES 6.8 to 8.0 and if there are improvments in ES, we should try to make them available as soon as possible.

File: libbeat/idxmgmt/ilm/config.go [hunk 1]
```diff
@@ -70,8 +70,8 @@ var DefaultPolicy = common.MapStr{
 			"hot": common.MapStr{
 				"actions": common.MapStr{
 					"rollover": common.MapStr{
-						"max_size": "50gb",
-						"max_age":  "30d",
+						"max_primary_shard_size": "50gb",
+						"max_age":                "30d",
```
Comment (urso) [inline libbeat/idxmgmt/ilm/config.go:74]: urgh... MapStr and global vars. We should have introduced a function that returns the default policy. Can you change it? The function would also get the ES version as parameter. The ES version is already known to the index manager, which is called by the ES connection callback.

Comment (ph) [reply libbeat/idxmgmt/ilm/config.go:74]: Make sense I will make the change here.

**Comment by EricDavisX** [comment]
@ruflin and @elastic/agent is this still relevant and valuable enough to pick up and finish off?

**Comment by ruflin** [comment]
I think the change still makes sense but likely not to priority at the moment.

**Comment by ph** [comment]
@jpountz is that still relevent?

**Comment by dakrone** [comment]
This is still relevant, and should be done. As we move more to an ingest-autoscaling design, we want to be able to increase the number of primary shards for an index automatically without having to change the ILM policy. This change makes the ILM policy independent of the number of shards.

**Comment by jlind23** [comment]
@ph any update here?

**Comment by ph** [comment]
I haven't made progress on this one, yes.

**Comment by ruflin** [comment]
@ph Can you push this over the line? As we now only support newer versions of Elasticsearch, this change should be pretty simple and no need to do more refactoring.

**Comment by ph** [comment]
@jlind23 @kvch I've rebased this on top of master.

**Comment by jlind23** [comment]
@ph it seems like a bunch of unit tests were relying on the max_size key. Shouldn't we update them?

File: libbeat/idxmgmt/ilm/config.go (hunk 1 above)
Comment (rdner) [reply libbeat/idxmgmt/ilm/config.go:52]: @ph I don't think you pushed that change yet but this thread got resolved.

Comment (ph) [reply libbeat/idxmgmt/ilm/config.go:52]: @rdner good point, I should have added a comment, I would prefer we push this change in, In the current state of beats and the move to Elastic-Agent I don't think we should add more time to remove the global vars and dealing with MapStr.

**Review by jlind23 — APPROVED** [review]

**Review by rdner — APPROVED** [review]

**Pull request merged**